articles/build-a-restful-api-with-adonisjs/index.md

```diff
@@ -0,0 +1,419 @@
+
+Cheers!
+
+### Creating Models
+Now, we will create all the models we need this API and configure them properly to interact with our database.
+
+```bash
+    node ace make:model Forum
+    node ace make:model Post
+```
+
+You can [clone my repository](https://github.com/Kaperskyguru/adonisjs-forum-api) to see how we map the columns and configure the relationships too.
+
+This is a preview of how the Model looks like for a `User` model:
+
+```ts
+import { DateTime } from "luxon";
+import Post from "App/Models/Post";
+import Forum from "App/Models/Forum";
+import Hash from "@ioc:Adonis/Core/Hash";
+import {
+  column,
+  beforeSave,
+  BaseModel,
+  hasMany,
+  HasMany,
+} from "@ioc:Adonis/Lucid/Orm";
+export default class User extends BaseModel {
+  @column({ isPrimary: true })
+  public id: number;
+  @column()
+  public email: string;
+  @column()
+  public name: string;
+  @column({ serializeAs: null })
+  public password: string;
+  @column()
+  public rememberMeToken?: string;
+  @column.dateTime({ autoCreate: true })
+  public createdAt: DateTime;
+  @column.dateTime({ autoCreate: true, autoUpdate: true })
+  public updatedAt: DateTime;
+  @hasMany(() => Post)
+  public posts: HasMany<typeof Post>;
+  @hasMany(() => Forum)
+  public forums: HasMany<typeof Forum>;
+  @beforeSave()
+  public static async hashPassword(user: User) {
+    if (user.$dirty.password) {
+      user.password = await Hash.make(user.password);
+    }
+  }
+}
+```
+
+### Creating Controllers
+Next, we are going to create the individual controllers for each Model that we have created above but before that, let’s create an AuthController for the authentication.
+
+```bash
+    node ace make:controller Auth
+```
+
+Open the file in `app/Controllers/Http/AuthController.ts` and paste the codes below.
+
+```ts
+import { HttpContextContract } from "@ioc:Adonis/Core/HttpContext";
+import User from "App/Models/User";
+
+export default class AuthController {
+  public async login({ request, auth }: HttpContextContract) {
+    const email = request.input("email");
+    const password = request.input("password");
+    const token = await auth.use("api").attempt(email, password, {
+      expiresIn: "10 days",
+    });
+    return token.toJSON();
+  }
+  public async register({ request, auth }: HttpContextContract) {
+    const email = request.input("email");
+    const password = request.input("password");
+    const name = request.input("name");
+    const newUser = new User();
+    newUser.email = email;
+    newUser.password = password;
+    newUser.name = name;
+    await newUser.save();
+    const token = await auth.use("api").login(newUser, {
+      expiresIn: "10 days",
+    });
+    return token.toJSON();
+  }
+}
+```
+
+The code above simply register and log in a user with any complex validations and error handling, so it can be easy to understand.
```

Comment:
should it be `registers and logins in a user`?

Reply:
done